File: sm_app/lib/controller/post_controller.dart
// ignore_for_file: prefer_final_fields, use_build_context_synchronously

import 'dart:io';
import 'package:dio/dio.dart' as dio;
import 'package:flutter/material.dart';
import 'package:get/get.dart';
import 'package:sm_app/model/post_model.dart';
import 'package:sm_app/network/network_api_service.dart';
import 'package:sm_app/utils/secure_storage.dart';
import 'package:sm_app/utils/urls.dart';
import 'package:sm_app/utils/utils.dart';

class PostController extends GetxController {
  SecureStorage _secureStorage = SecureStorage();
  final _apiService = NetworkApiService();

  Rx<PostModel?> _postModel = Rx<PostModel?>(null);
  get postModel => _postModel.value;

  Future<void> uploadPost(File imageFile, String title, String description,
      BuildContext context) async {
    try {
      String fileName = imageFile.path.split('/').last;

      String? uid = await _secureStorage.readStorageData('uid');
      if (uid == null) {
        throw Exception("User ID (uid) is missing");
      }

      dio.FormData formData = dio.FormData.fromMap({
        'post-image': await dio.MultipartFile.fromFile(imageFile.path,
            filename: fileName),
        'title': title,
        'description': description,
        'uid': uid,
      });

      dio.Response response = await dio.Dio().post(
        Urls.uploadPost,
        data: formData,
        options: dio.Options(headers: {'Content-Type': 'multipart/form-data'}),
      );
      Utils.displayMessages(response.data, context);
    } catch (e) {
      rethrow;
    }
  }

  Future<void> getPosts(BuildContext context) async {
    try {
      // Fetch the posts from the API
      final res = await _apiService.getGetApiResponse(Urls.getPosts);

      // Check if the response contains data
      if (res is Map<String, dynamic> && res.containsKey('data')) {
        // Create a new PostModel instance from the response
        _postModel.value = PostModel.fromJson(res);

        // Display success message using the string message
        if (res['msg'] != null) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text(res['msg'].toString())),
          );
        }
      } else {
        throw Exception('Invalid response format');
      }
    } catch (e) {
      print("Error fetching posts: $e");
      // Display error message
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Failed to fetch posts')),
      );
      rethrow;
    }
  }

  Future<void> likePost(BuildContext context, String postId) async {
    try {
      // Get the user ID from secure storage
      String? uid = await _secureStorage.readStorageData('uid');

      if (uid == null) {
        throw Exception('User ID not found');
      }
      print('Uid: $uid');
      print('postId $postId');
      // Prepare the data to send to the backend
      final Map data = {
        'postId': postId,
        'uid': uid,
      };

      // Make the API call to like the post
      final res = await _apiService.getPostApiResponse(
        Urls.likePost, // Make sure to add this endpoint in your Urls class
        data,
      );
      print(res);

      // Handle the response
      Utils.displayMessages(res, context);

      // If the response indicates success, you might want to update the local post model
      // This will depend on your backend response structure
      if (res['success'] == true) {
        // Optionally update the local post model to reflect the like
        // You might want to handle this based on your backend response
        await getPosts(
            context); // Refresh the posts to show updated like status
      }
    } catch (e) {
      print('Error in likePost: $e');

      rethrow; // Rethrow the error so it can be handled by the calling widget
    }
  }
}

File: backend/controllers/postController.js
import { Post } from "../models/postModel.js";
import { User } from "../models/userModel.js";

export const uploadPost = async (req, res) => {
  try {
    const { uid, title, description } = req.body;
    const filename = req.file.filename;

    const newPost = new Post({
      title: title,
      description: description,
      image: filename,
      user: uid,
    });

    // Save the post
    const savedPost = await newPost.save();

    if (savedPost) {
      return res.json({
        msg: "Post created successfully",
        status: "success",
      });
    } else {
      return res.json({
        msg: "Post creation failed",
        status: "error",
      });
    }
  } catch (error) {
    return res.json({ msg: error.message, status: "error" });
  }
};

export const getPosts = async (req, res) => {
  try {
    const posts = await Post.find({ isDel: false });
    return res.json({
      data: posts,
      msg: "posts fetched successfully",
      status: "success",
    });
  } catch (error) {
    return res.json({ msg: error.message, status: "error" });
  }
};

export const delPost = async (req, res) => {
  try {
    const { uid, postId } = req.body;

    if (!uid || !postId) {
      return res.json({ msg: "Missing required fields", status: "error" });
    }

    const [user, post] = await Promise.all([
      User.findById(uid),
      Post.findById(postId),
    ]);

    if (!user) {
      return res.json({ msg: "User not found", status: "error" });
    }

    if (!post) {
      return res.json({ msg: "Post not found", status: "error" });
    }

    post.isDel = true;

    await post.save();

    return res.json({
      status: "success",
      msg: "Post deleted successfully",
    });
  } catch (error) {
    return res.json({ msg: "Internal server error", status: "error" });
  }
};

export const likePost = async (req, res) => {
  try {
    const { uid, postId } = req.body;
    console.log(uid);
    console.log(postId);
    if (!uid || !postId) {
      return res.json({ msg: "Missing required fields", status: "error" });
    }

    const [user, post] = await Promise.all([
      User.findById(uid),
      Post.findById(postId),
    ]);

    if (!user) {
      return res.json({ msg: "User not found", status: "error" });
    }

    if (!post) {
      return res.json({ msg: "Post not found", status: "error" });
    }

    const hasLiked = post.likes.includes(uid);

    if (hasLiked) {
      post.likes = post.likes.filter((id) => id.toString() !== uid.toString());
    } else {
      post.likes.push(uid);
    }

    await post.save();

    return res.json({
      status: "success",
      msg: hasLiked ? "Post unliked successfully" : "Post liked successfully",
      likes: post.likes,
    });
  } catch (error) {
    console.error("Like API Error:", error);
    return res.json({ msg: "Internal server error", status: "error" });
  }
};
